### PR TITLE
releases: add `release version` property

### PR DESCRIPTION
**Pull request by xginn8**
Connects-to: https://github.com/balena-io/open-balena-api/issues/378
HQ: https://github.com/balena-io/balena-io/issues/2219
Change-type: minor
Signed-off-by: Matthew McGinn <matthew@balena.io>

### CODE REVIEW

**Comment by xginn8** [comment]
I implemented the uniqueness constraint outside of SBVR following how app name uniqueness is validated (https://github.com/balena-io/open-balena-api/blob/master/src/hooks/resources/application.ts#L74-L99).

**Comment by xginn8** [comment]
@balena-ci rebase

File: src/hooks/resources/release.ts [hunk 1]
```diff
@@ -127,12 +131,87 @@ sbvrUtils.addPureHook('POST', 'resin', 'release', {
...
+			options: {
+				$select: 'belongs_to__application',
+				$filter: {
+					id: { $in: releaseIds },
+				},
+			},
+		})) as AnyObject[];
+		if (
+			request.values.release_name !== undefined &&
+			request.values.release_name != null
```
Comment (CameronDiver) [inline src/hooks/resources/release.ts:152]: you can drop the `!== undefined` check, as `!= null` and `== null` will coerce undefined to null and behave correctly.

File: src/hooks/resources/release.ts [hunk 2]
```diff
@@ -127,12 +131,87 @@ sbvrUtils.addPureHook('POST', 'resin', 'release', {
...
+			})) as AnyObject[];
+			if (conflictingValidNamedReleases.length !== 0) {
+				throw new ConflictError(
+					'Each valid release in an application must be named uniquely',
+				);
+			}
+		}
+		if (
+			request.values.release_version !== undefined &&
+			request.values.release_version != null
```
Comment (CameronDiver) [inline src/hooks/resources/release.ts:179]: here too

File: src/hooks/resources/release.ts [hunk 3]
```diff
@@ -127,12 +131,87 @@ sbvrUtils.addPureHook('POST', 'resin', 'release', {
...
+			const conflictingValidNamedReleases = (await args.api.get({
+				resource: 'release',
+				options: {
+					$select: 'id',
+					$filter: {
+						belongs_to__application: {
+							$in: _.uniq(
+								_.map(relIds, (rel) => {
+									return rel.belongs_to__application.__id;
+								}),
```
Comment (CameronDiver) [inline src/hooks/resources/release.ts:163]:
```suggestion
								_.map(relIds, 'belongs_to__application.__id')
```

File: src/hooks/resources/release.ts [hunk 4]
```diff
@@ -127,12 +131,87 @@ sbvrUtils.addPureHook('POST', 'resin', 'release', {
...
+			const conflictingValidVersionedReleases = (await args.api.get({
+				resource: 'release',
+				options: {
+					$select: 'id',
+					$filter: {
+						belongs_to__application: {
+							$in: _.uniq(
+								_.map(relIds, (rel) => {
+									return rel.belongs_to__application.__id;
+								}),
```
Comment (CameronDiver) [inline src/hooks/resources/release.ts:190]:
```suggestion
								_.map(relIds, 'belongs_to__application.__id')
```

File: src/hooks/resources/release.ts (hunk 1 above)
Comment (xginn8) [reply src/hooks/resources/release.ts]: done

File: src/hooks/resources/release.ts (hunk 2 above)
Comment (xginn8) [reply src/hooks/resources/release.ts]: done

**Comment by xginn8** [comment]
@balena-ci rebase

**Comment by xginn8** [comment]
This change just lacks the following complex term from https://www.flowdock.com/app/rulemotion/resin-tech/threads/1KMxrVlGHKqIaDbVJYGqea0f0VR, as I work on getting the syntax right:

> I think the ideal form would be more like:

```
    Term: active release
    Definition: a release that has a status that is equal to "success" and is not invalidated
    Rule: It is necessary that each active release that has a release version1, belongs to an application that owns exactly one active release that has a release version that is equal to the release version1. 
```

File: test/test-lib/fixtures.ts [hunk 5]
```diff
@@ -98,6 +98,37 @@ const loaders: Dictionary<LoaderFunc> = {
...
+			logErrorAndThrow(`Could not find application: ${jsonData.application}`);
+		}
+
+		return createResource({
+			resource: 'release',
+			body: {
+				belongs_to__application: application.id,
+				belongs_to__user: user.id,
+				start_timestamp: Date.now(),
+				end_timestamp: Date.now(),
```
Comment (Page-) [inline test/test-lib/fixtures.ts:118]: Did you want these to be identical or is a potentially ms difference fine?

Comment (xginn8) [reply test/test-lib/fixtures.ts:118]: I don't see this as a problem, nothing strictly depends on time here.

**Pull request merged**